modules/currency-l0/src/main/scala/io/constellationnetwork/currency/l0/snapshot/services/StateChannelBinarySender.scala

```diff
@@ -82,59 +103,75 @@ object StateChannelBinarySender {
-            state.pending.splitAt(cutAt + 1)._2
+          val indexedTracked = state.tracked.zipWithIndex
+
+          val maybeHighestConfirmationIndex = indexedTracked.collect {
+            case (PendingBinary(tracked, _, _), index) if confirmedHashesInGlobalSnapshot.contains(tracked.hash) => index
+          }.maxOption
+
+          val updatedTracked = indexedTracked.map {
+            case (pendingBinary @ PendingBinary(tracked, enqueuedAtOrdinal, sendsSoFar), index)
+                if index <= maybeHighestConfirmationIndex.getOrElse(-1) =>
+              ConfirmedBinary(pendingBinary, GlobalSnapshotConfirmationProof.fromGlobalSnapshot(globalSnapshot))
+            case (other, _) => other
           }
 
           val updatedRetryMode = {
             val hasStalled =
-              updatedPending.exists(p => globalSnapshot.ordinal.value - p.enqueuedAtOrdinal.value >= noConfirmationsToTriggerRetryMode)
+              updatedTracked.exists {
+                case PendingBinary(_, enqueuedAtOrdinal, _) =>
+                  globalSnapshot.ordinal.value - enqueuedAtOrdinal.value >= noConfirmationsToTriggerRetryMode
+                case _ => false
+              }
 
             if (!state.retryMode) {
               hasStalled
-            } else if (state.cap >= state.pending.length && updatedPending.forall(_.sendsSoFar >= 1) && !hasStalled) {
-              false
-            } else
-              state.retryMode
+            } else {
+              val pendingCount = state.tracked.collect { case _: PendingBinary => 1 }.sum
+              val allPendingAlreadySent = updatedTracked.forall {
+                case PendingBinary(_, _, sendsSoFar) => sendsSoFar >= 1
+                case _                               => true
+              }
+              if (pendingCount <= state.cap && allPendingAlreadySent && !hasStalled) {
+                false
+              } else
+                state.retryMode
+            }
           }
 
           val (updatedCap, updatedBackoffExponent, updatedNoConfirmationsSinceRetryCount) =
-            if ((!updatedRetryMode && state.retryMode) || updatedPending.isEmpty) {
+            if ((!updatedRetryMode && state.retryMode) || updatedTracked.isEmpty) {
               val empty = State.empty
               val cap = empty.cap
               val backoffExponent = empty.backoffExponent
               val noConfirmationsSinceRetryCount = empty.noConfirmationsSinceRetryCount
 
               (cap, backoffExponent, noConfirmationsSinceRetryCount)
             } else if (updatedRetryMode) {
-              val confirmedCount = maybeConfirmationIndex.map(_ + 1L).getOrElse(0L)
+              val confirmedCount = maybeHighestConfirmationIndex.map(_ + 1L).getOrElse(0L)
 
               if (confirmedCount > 0) {
                 val cap = {
                   val maxCap = confirmedCount * confirmedCountMultiplier
                   val log2 = (x: Double) => Math.log10(x) / Math.log10(2.0)
```

Comment:
Alternative log2 operation if you want to avoid the floating point operation, it is recursive though so maybe the floating point is more performant.

        @tailrec
        def log2(n: Int, acc: Int = 0): Int =
          if (n <= 1) acc
          else log2(n >> 1, acc + 1)

Reply:
In floating point based solution `log2(9)` is 3.169925001442312 so `Math.ceil` makes it 4. In your solution it will be 3. Since both functions produce different results I would leave it as it is for now to not change the maths behind the retry mode. I do not remember if `Math.ceil` was used on purpose so I would rather keep it as it is for now.